import grpc
from concurrent import futures
import time
from django.core.management.base import BaseCommand

from rules import profile_pb2_grpc
from rules import rule_config_pb2
from google.protobuf import empty_pb2

class ProfileService(profile_pb2_grpc.ProfileServiceServicer):
    def ProcessProfileStream(self, request_iterator, context):
        """
        –≠—Ç–æ—Ç –º–µ—Ç–æ–¥ –±—É–¥–µ—Ç –≤—ã–∑–≤–∞–Ω, –∫–æ–≥–¥–∞ Django-–∫–ª–∏–µ–Ω—Ç –æ—Ç–ø—Ä–∞–≤–∏—Ç —Å—Ç—Ä–∏–º –ø—Ä–æ—Ñ–∏–ª–µ–π.
        'request_iterator' - —ç—Ç–æ –∏—Ç–µ—Ä–∞—Ç–æ—Ä –æ–±—ä–µ–∫—Ç–æ–≤ Profile.
        """
        print("‚úÖ === –ü–æ–ª—É—á–µ–Ω —Å—Ç—Ä–∏–º –ø—Ä–æ—Ñ–∏–ª–µ–π –æ—Ç Django Admin! === ‚úÖ")
        
        try:
            profiles_received = 0
            
            for profile in request_iterator:
                profiles_received += 1
                print(f"\n--- –ü—Ä–æ—Ñ–∏–ª—å ---")
                print(f"  UUID: {profile.uuid}")
                print(f"  Name: {profile.name}")
                print(f"  –ö–æ–ª–∏—á–µ—Å—Ç–≤–æ –ø—Ä–∞–≤–∏–ª: {len(profile.rules)}")
                
                for rule_config in profile.rules:
                    print(f"    - –ö–æ–Ω—Ñ–∏–≥ '{rule_config.name}' (UUID: {rule_config.uuid})")
                    print(f"      –¢–∏–ø –ø—Ä–∞–≤–∏–ª–∞: {rule_config_pb2.RuleConfig.RuleType.Name(rule_config.rule_type)}")
                    print(f"      –ö—Ä–∏—Ç–∏—á–Ω–æ—Å—Ç—å: {rule_config.is_critical}")

                    active_rule_field = rule_config.WhichOneof('rule')

                    if active_rule_field:
                        rule_object = getattr(rule_config, active_rule_field)
                        
                        from google.protobuf import json_format
                        rule_json = json_format.MessageToDict(rule_object)
                        print(f"      –°–æ–¥–µ—Ä–∂–∏–º–æ–µ –ø—Ä–∞–≤–∏–ª–∞ (JSON): {rule_json}")
                    else:
                        print("      –°–æ–¥–µ—Ä–∂–∏–º–æ–µ –ø—Ä–∞–≤–∏–ª–∞: (–Ω–µ –∑–∞–¥–∞–Ω–æ)")
            
            print(f"\n‚úÖ === –û–±—Ä–∞–±–æ—Ç–∞–Ω–æ {profiles_received} –ø—Ä–æ—Ñ–∏–ª–µ–π! === ‚úÖ")
            
            return empty_pb2.Empty()
        except Exception as e:
            print(f"‚ùå –û—à–∏–±–∫–∞ –ø—Ä–∏ –æ–±—Ä–∞–±–æ—Ç–∫–µ –∑–∞–ø—Ä–æ—Å–∞: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(str(e))
            return empty_pb2.Empty()

class Command(BaseCommand):
    help = '–ó–∞–ø—É—Å–∫–∞–µ—Ç gRPC –º–æ–∫-—Å–µ—Ä–≤–µ—Ä –¥–ª—è —Å–µ—Ä–≤–∏—Å–∞ Director'

    def handle(self, *args, **options):
        server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
        profile_pb2_grpc.add_ProfileServiceServicer_to_server(ProfileService(), server)
        
        port = "50051"
        server.add_insecure_port(f"[::]:{port}")
        
        server.start()
        self.stdout.write(self.style.SUCCESS(f"üöÄ –ú–æ–∫-—Å–µ—Ä–≤–µ—Ä Director –∑–∞–ø—É—â–µ–Ω –Ω–∞ –ø–æ—Ä—Ç—É {port}..."))
        
        try:
            server.wait_for_termination()
        except KeyboardInterrupt:
            self.stdout.write(self.style.WARNING("üõë –°–µ—Ä–≤–µ—Ä –æ—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ—Ç—Å—è."))
            server.stop(0)